File: lab5_13_1/1.py
import heapq
import sys

class Node:
    def __init__(self, node_id, h_cost):
        self.id = node_id
        self.h_cost = h_cost

    def __lt__(self, other):
        return self.h_cost < other.h_cost

def best_first_search(adj, start, target, h_values):
    # Min-priority queue: stores Node objects
    pq = [Node(start, h_values[start])]
    
    visited = {start: True}
    parent = {start: None}
    expanded_nodes = []

    found = False

    while pq:
        current = heapq.heappop(pq)
        u = current.id

        expanded_nodes.append(u)

        if u == target:
            found = True
            break

        # Check neighbors
        if u in adj:
            for v in adj[u]:
                if v not in visited:
                    visited[v] = True
                    parent[v] = u
                    heapq.heappush(pq, Node(v, h_values[v]))

    print("\nRESULT")
    print("Expanded Nodes: " + " -> ".join(expanded_nodes))

    if found:
        print("Path Found: ", end="")
        path = []
        crawl = target
        while crawl is not None:
            path.append(crawl)
            crawl = parent[crawl]
        path.reverse()
        print(" -> ".join(path))
    else:
        print("Path Found: No path exists.")

def main():
    try:
        print("Enter number of Nodes and Edges: ", end="", flush=True)
        line1 = sys.stdin.readline().split()
        if not line1: return
        n = int(line1[0])
        m = int(line1[1])

        adj = {}
        print("Enter Edges (Source Destination) :")
        for _ in range(m):
            edge = sys.stdin.readline().split()
            if not edge: break
            u, v = edge[0], edge[1]
            if u not in adj: adj[u] = []
            adj[u].append(v)

        print("Enter Start Node and Goal Node: ", end="", flush=True)
        line2 = sys.stdin.readline().split()
        if not line2: return
        start, target = line2[0], line2[1]

        print(f"Enter Heuristic value for {n} nodes (Node Value) :")
        h_values = {}
        for _ in range(n):
            h_line = sys.stdin.readline().split()
            if not h_line: break
            node, val = h_line[0], int(h_line[1])
            h_values[node] = val

        best_first_search(adj, start, target, h_values)
        print("23052247\nKush Singh\nCSE-22")
    except (ValueError, EOFError, IndexError):
        pass

if __name__ == "__main__":
    main()

File: lab5_13_1/1.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <stack>
#include <algorithm>

using namespace std;

struct Node {
    char id;
    int h_cost;

    Node(char i, int h) : id(i), h_cost(h) {}

    bool operator >(const Node& other) const {
        return h_cost > other.h_cost;
    }
};

void bestFirstSearch(map<char, vector<char>>& adj, char start, char target, map<char, int>& h_values) {
    priority_queue<Node, vector<Node>, greater<Node>> pq;
    
    map<char, bool> visited;
    map<char, char> parent;
    vector<char> expanded_nodes; 

    // Initialize
    pq.push(Node(start, h_values[start]));
    visited[start] = true;
    parent[start] = '\0'; 

    bool found = false;

    while (!pq.empty()) {
        Node current = pq.top();
        pq.pop();
        char u = current.id;

        expanded_nodes.push_back(u);

        if (u == target) {
            found = true;
            break;
        }

        // Check neighbors
        for (char v : adj[u]) {
            if (!visited[v]) {
                visited[v] = true;
                parent[v] = u;
                pq.push(Node(v, h_values[v]));
            }
        }
    }

    cout << "\nRESULT" << endl;
    cout << "Expanded Nodes: ";
    for (size_t i = 0; i < expanded_nodes.size(); i++) {
        cout << expanded_nodes[i];
        if (i < expanded_nodes.size() - 1) cout << " -> ";
    }
    cout << endl;

    if (found) {
        cout << "Path Found: ";
        stack<char> path;
        char crawl = target;
        while (crawl != '\0') {
            path.push(crawl);
            crawl = parent[crawl];
        }

        while (!path.empty()) {
            cout << path.top();
            path.pop();
            if (!path.empty()) cout << " -> ";
        }
        cout << endl;
    } else {
        cout << "Path Found: No path exists." << endl;
    }
}

int main() {
    int n, m;
    map<char, vector<char>> adj;
    map<char, int> h_values;

    cout << "Enter number of Nodes and Edges: ";
    if (!(cin >> n >> m)) return 0;

    cout << "Enter Edges (Source Destination) :\n";
    for(int i = 0; i < m; i++) {
        char u, v;
        cin >> u >> v;
        adj[u].push_back(v); 
    }

    char start, target;
    cout << "Enter Start Node and Goal Node: ";
    cin >> start >> target;

    cout << "Enter Heuristic value for " << n << " nodes (Node Value) :\n";
    for(int i = 0; i < n; i++) {
        char node;
        int val;
        cin >> node >> val;
        h_values[node] = val;
    }

    bestFirstSearch(adj, start, target, h_values);
    
    cout << "23052247\nKush Singh\nCSE-22\n";
    return 0;
}

/*
    Output :
    Enter number of Nodes and Edges: 4 4
    Enter Edges (Source Destination) :
    A B
    A C
    B D
    C D
    Enter Start Node and Goal Node: A D
    Enter Heuristic value for 4 nodes (Node Value) e.g., W 10:
    A 5
    B 3
    C 4
    D 0
    
    RESULT
    Expanded Nodes: A -> B -> D
    Path Found: A -> B -> D
    23052247
    Kush Singh
    CSE-22
 */